src/dataflow/DataflowAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,379 @@
+        TypeElement typeElt = types.boxedClass(type.getUnderlyingType());
+        AnnotationMirror am = DataflowUtils.createDataflowAnnotation(typeElt.asType().toString(),
+                this.processingEnv);
+        AnnotatedDeclaredType dt = fromElement(typeElt);
+        dt.addAnnotation(am);
+        return dt;
+    }
+
+    /**
+     * This method handles unboxing for reference type.
+     * For statements, int i = new Integer(3);
+     * The annotation for i should be @DataFlow(typeNames = {"int"}).
+     */
+    @Override
+    public AnnotatedPrimitiveType getUnboxedType(AnnotatedDeclaredType type)
+            throws IllegalArgumentException {
+        PrimitiveType primitiveType = types.unboxedType(type.getUnderlyingType());
+        AnnotationMirror am = DataflowUtils.createDataflowAnnotation(primitiveType.toString(),
+                this.processingEnv);
+        AnnotatedPrimitiveType pt = (AnnotatedPrimitiveType) AnnotatedTypeMirror.createType(
+                primitiveType, this, false);
+        pt.addAnnotation(am);
+        return pt;
+    }
+
+    private final class DataFlowQualifierHierarchy extends GraphQualifierHierarchy {
+
+        public DataFlowQualifierHierarchy(MultiGraphFactory f, AnnotationMirror bottom) {
+            super(f, bottom);
+        }
+
+        /**
+         * This method checks whether rhs is subtype of lhs. rhs and lhs are
+         * both Dataflow types with typeNameRoots argument.
+         * 
+         * @param rhs
+         * @param lhs
+         * @return true is rhs is subtype of lhs, otherwise return false.
+         */
+        private boolean isSubtypeWithRoots(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            Set<String> rTypeNamesSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNames(rhs)));
+            Set<String> lTypeNamesSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNames(lhs)));
+            Set<String> rRootsSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNameRoots(rhs)));
+            Set<String> lRootsSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNameRoots(lhs)));
+            Set<String> combinedTypeNames = new HashSet<String>();
+            combinedTypeNames.addAll(rTypeNamesSet);
+            combinedTypeNames.addAll(lTypeNamesSet);
+            Set<String> combinedRoots = new HashSet<String>();
+            combinedRoots.addAll(rRootsSet);
+            combinedRoots.addAll(lRootsSet);
+
+            AnnotationMirror combinedAnno = DataflowUtils.createDataflowAnnotationWithRoots(
+                    combinedTypeNames, combinedRoots, processingEnv);
+            AnnotationMirror refinedCombinedAnno = refineDataflow(combinedAnno);
+            AnnotationMirror refinedLhs = refineDataflow(lhs);
+
+            if (AnnotationUtils.areSame(refinedCombinedAnno, refinedLhs)) {
```

Comment:
Umm, it makes sense. Thanks for detailed explanation! It really helps! Now I understand dataflow system better.